segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
     }
 
     /**
-     * Gets a value indicating how much to wait for the service to shut down, before failing it.
+     * Adds given chunks to list of garbage chunks.
      *
-     * @return The Duration.
+     * @param chunksToDelete List of chunks to delete.
      */
-    @Override
-    protected Duration getShutdownTimeout() {
-        return SHUTDOWN_TIMEOUT;
+    CompletableFuture<Void> addChunksToGarbage(long transactionId, Collection<String> chunksToDelete) {
+        if (null != taskQueue) {
+            val currentTime = currentTimeSupplier.get();
+            val futures = new ArrayList<CompletableFuture<Void>>();
+            chunksToDelete.forEach(chunkToDelete -> futures.add(addChunkToGarbage(transactionId, chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0)));
+            return Futures.allOf(futures);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
     /**
-     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     * Adds given chunk to list of garbage chunks.
      *
-     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
-     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     * @param chunkToDelete Name of the chunk to delete.
+     * @param startTime Start time.
+     * @param attempts Number of attempts to delete this chunk so far.
      */
-    @Override
-    protected CompletableFuture<Void> doRun() {
-        loopFuture = Futures.loop(
-                this::canRun,
-                () -> delaySupplier.get()
-                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionMaxConcurrency()), executor)
-                        .handleAsync((v, ex) -> {
-                            if (null != ex) {
-                                log.error("{}: Error during doRun.", traceObjectId, ex);
-                            }
-                            return null;
-                        }, executor),
-                executor);
-        return loopFuture;
+    CompletableFuture<Void>  addChunkToGarbage(long transactionId, String chunkToDelete, long startTime, int attempts) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunkToDelete, startTime, attempts, TaskInfo.DELETE_CHUNK, transactionId))
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_CHUNK_QUEUED.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
-    private boolean canRun() {
-        return isRunning() && getStopException() == null && !closed.get();
+    CompletableFuture<Void>  addSegmentToGarbage(long transactionId, String segmentToDelete) {
+        if (null != taskQueue) {
```

Comment:
If `taskQueue`, this method just does nothing right? Is that ok or we should check that `taskQueue` is not null and throw otherwise?

Reply:
It is checked during initialization of container and as such it is not possible to have.
However during many of unit tests GC is not initialized and the taskQueue will be null and in fact it is not required to be initialized in that particular scenario. 
In that sense it is okay for it to be null in unit tests and yet it is correctly enforced when GC tests initialize GC and integration tests initialize container etc.